File: services/ai/schemas/validator.ts
import Ajv from 'ajv';
import reflectionSchema from './reflection.schema.json';
import planSchema from './plan.schema.json';
import scheduleSchema from './schedule.schema.json';

const ajv = new Ajv({ strict: true, allErrors: true });

const compiledSchemas = {
  reflection: ajv.compile(reflectionSchema.schema),
  plan: ajv.compile(planSchema.schema),
  schedule: ajv.compile(scheduleSchema.schema),
};

export type SchemaType = keyof typeof compiledSchemas;

export interface ValidationResult {
  valid: boolean;
  errors?: string[];
}

export function validateAgainstSchema(type: SchemaType, data: any): ValidationResult {
  const validate = compiledSchemas[type];
  const valid = validate(data);
  
  if (valid) {
    return { valid: true };
  } else {
    return {
      valid: false,
      errors: validate.errors?.map(err => `${err.instancePath} ${err.message}`) || ['Unknown validation error']
    };
  }
}

// Retry logic for model calls
export async function callWithRetry<T>(
  callFn: (retryHint?: string) => Promise<T>,
  maxRetries: number = 2,
  validator?: (result: T) => ValidationResult
): Promise<T> {
  let lastError: string | undefined;
  
  for (let attempt = 0; attempt <= maxRetries; attempt++) {
    try {
      const result = await callFn(lastError);
      
      if (validator) {
        const validation = validator(result);
        if (!validation.valid) {
          lastError = `Validation failed: ${validation.errors?.join(', ')}`;
          if (attempt < maxRetries) continue;
          throw new Error(lastError);
        }
      }
      
      return result;
    } catch (error) {
      lastError = error instanceof Error ? error.message : 'Unknown error';
      if (attempt >= maxRetries) {
        throw new Error(`Failed after ${maxRetries + 1} attempts. Last error: ${lastError}`);
      }
    }
  }
  
  throw new Error('Unexpected error in retry logic');
}

File: services/python-ai/app/ai_legacy/schemas/validator.py
from typing import Any, Callable, Dict, List, Literal, Optional

import jsonschema
import json
from pathlib import Path

SchemaType = Literal["reflection", "plan", "schedule"]


def _load_schema(name: str) -> Dict[str, Any]:
    schema_path = Path(__file__).with_name(f"{name}.schema.json")
    with schema_path.open("r", encoding="utf-8") as f:
        return json.load(f)["schema"]


_compiled_schemas = {
    "reflection": jsonschema.Draft7Validator(_load_schema("reflection")),
    "plan": jsonschema.Draft7Validator(_load_schema("plan")),
    "schedule": jsonschema.Draft7Validator(_load_schema("schedule")),
}


def validate_against_schema(schema_type: SchemaType, data: Any) -> Dict[str, Any]:
    validator = _compiled_schemas[schema_type]
    errors = sorted(validator.iter_errors(data), key=lambda e: e.path)
    if not errors:
        return {"valid": True}
    return {"valid": False, "errors": [f"{'/'.join(map(str, err.path))} {err.message}" for err in errors]}


async def call_with_retry(call_fn: Callable[[Optional[str]], Any], max_retries: int = 2, validator: Optional[Callable[[Any], Dict[str, Any]]] = None) -> Any:
    last_error: Optional[str] = None
    for attempt in range(max_retries + 1):
        try:
            result = await call_fn(last_error)
            if validator:
                validation = validator(result)
                if not validation.get("valid"):
                    last_error = f"Validation failed: {', '.join(validation.get('errors', []))}"
                    if attempt < max_retries:
                        continue
                    raise ValueError(last_error)
            return result
        except Exception as error:  # pylint: disable=broad-except
            last_error = str(error)
            if attempt >= max_retries:
                raise RuntimeError(f"Failed after {max_retries + 1} attempts. Last error: {last_error}") from error
    raise RuntimeError("Unexpected error in retry logic")

